File: Source/Remote Control Server/UDP.vb
Imports System
Imports System.IO
Imports System.Net
Imports System.Net.Sockets
Imports System.Windows.Threading
Imports System.Threading

Public Class UDP

    Public Const port As Integer = 1926
    Public Const timeout As Integer = 2000

    Public udpClient As UdpClient

    Public isListening As Boolean = False
    Private listenThread As Thread

    Public Sub initialize()
        Logger.add("Initializing UDP")

        startListener()
    End Sub

    Public Sub sendData(ByRef command As Command)

    End Sub

    Public Sub startListener()
        Try
            If Not isListening Then
                isListening = True

                udpClient = New UdpClient(port)
                udpClient.Client.ReceiveTimeout = timeout

                listenThread = New Thread(New ThreadStart(AddressOf listen))
                listenThread.Start()

                Logger.add("UDP listener started at port " & port)
            Else
                Logger.add("UDP listener already running at port " & port)
            End If
        Catch ex As Exception
            isListening = False
            Logger.add("Error while starting UDP listener:")
            Logger.add(ex.ToString)
        End Try
    End Sub

    Public Sub stopListener()
        isListening = False
    End Sub

    Public Sub listen()
        Do While isListening
            Try
                Dim message_data As Byte()
                message_data = udpClient.Receive(Network.anyIpEndPoint)
                Try
                    If Not message_data Is Nothing Then
                        Dim command As Command = New Command
                        command.source = Remote.lastCommand.source
                        command.destination = Remote.lastCommand.destination
                        command.data = message_data
                        command.process()
                    End If
                Catch ex As Exception
                    Logger.add("Error while receiving UDP data:")
                    Logger.add(ex.ToString)
                End Try
            Catch ex As Exception
                'udpClient.Receive timed out, this is alright
            End Try
        Loop
    End Sub
End Class

File: Source/MacOS/RemoteControlServer/udp.cpp
#include "logger.h"
#include "remote.h"
#include "udp.h"

#include <QByteArray>
#include <QDebug>
#include <QTimer>

UDP* UDP::instance = NULL;

UDP* UDP::Instance()
{
    if (!instance)
    {
        instance = new UDP();
    }
    return instance;
}

UDP::UDP() :
    port(1926),
    timeout(2000)
{
    isListening = false;
    Logger::Instance()->add("Initializing UDP");

    // let's restart the listener every 15 seconds as it seems that it stops working after a certain amount of time or received data
    QTimer *timer = new QTimer();
    connect(timer, SIGNAL(timeout()), this, SLOT(restartListener()));
    timer->start(15000);

    startListener(true);
}

void UDP::restartListener()
{
    stopListener();
    startListener(false);
}

void UDP::startListener(bool logMessages)
{
    if (!isListening)
    {
        isListening = true;
        udpSocket = new QUdpSocket(this);
        bool success = udpSocket->bind(QHostAddress::Any, port);
        if (success)
        {
            connect(udpSocket, SIGNAL(readyRead()), this, SLOT(listen()));
            if (logMessages) Logger::Instance()->add("UDP listener started listening at port " + QString::number(port));
        }
        else
        {
            isListening = false;
            if (logMessages) Logger::Instance()->add("Error while starting UDP listener");
        }
    }
    else
    {
        if (logMessages) Logger::Instance()->add("UDP listener already running at port " + QString::number(port));
    }
}

void UDP::stopListener()
{
    isListening = false;
    udpSocket->close();
}

void UDP::listen()
{
    while (udpSocket->hasPendingDatagrams())
    {
        QByteArray *message_data = new QByteArray();
        message_data->resize(udpSocket->pendingDatagramSize());
        QHostAddress sender;
        quint16 senderPort;
        udpSocket->readDatagram(message_data->data(), message_data->size(), &sender, &senderPort);
        if (message_data->length() > 0)
        {
            Command *command = new Command();
            command->source = Remote::Instance()->lastCommand->source;
            command->destination = Remote::Instance()->lastCommand->destination;
            command->data = message_data;
            command->process();
        }
        else
        {
            Logger::Instance()->add("Error while receiving UDP data");
        }
    }
}
